q2_sample_classifier/tests/test_visualization.py

```diff
@@ -0,0 +1,224 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2017-2020, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+import pandas as pd
+import pandas.util.testing as pdt
+from os import mkdir
+from os.path import join
+import biom
+
+import qiime2
+from qiime2.plugins import sample_classifier, feature_table
+
+from q2_sample_classifier.visuals import (
+    _linear_regress, _calculate_baseline_accuracy,
+    _add_sample_size_to_xtick_labels)
+from q2_sample_classifier.classify import (
+    scatterplot, confusion_matrix)
+from q2_sample_classifier.utilities import (
+    _match_series_or_die, _predict_and_plot)
+from q2_sample_classifier.tests.test_base_class import \
+    SampleClassifierTestPluginBase
+
+
+class TestVisuals(SampleClassifierTestPluginBase):
+
+    md = pd.DataFrame([(1, 'a', 0.11), (1, 'a', 0.12), (1, 'a', 0.13),
+                       (2, 'a', 0.19), (2, 'a', 0.18), (2, 'a', 0.21),
+                       (1, 'b', 0.14), (1, 'b', 0.13), (1, 'b', 0.14),
+                       (2, 'b', 0.26), (2, 'b', 0.27), (2, 'b', 0.29)],
+                      columns=['Time', 'Group', 'Value'])
+
+    def test_linear_regress(self):
+        res = _linear_regress(self.md['Value'], self. md['Time'])
+        self.assertAlmostEqual(res.iloc[0]['Mean squared error'], 1.9413916666)
+        self.assertAlmostEqual(res.iloc[0]['r-value'], 0.86414956372460128)
+        self.assertAlmostEqual(res.iloc[0]['r-squared'], 0.74675446848541871)
+        self.assertAlmostEqual(res.iloc[0]['P-value'], 0.00028880275858705694)
+
+    def test_calculate_baseline_accuracy(self):
+        accuracy = 0.9
+        y_test = pd.Series(['a', 'a', 'a', 'b', 'b', 'b'], name="class")
+        classifier_accuracy = _calculate_baseline_accuracy(y_test, accuracy)
+        expected_results = (6, 3, 0.5, 1.8)
+        for i in zip(classifier_accuracy, expected_results):
+            self.assertEqual(i[0], i[1])
+
+
+class TestHeatmap(SampleClassifierTestPluginBase):
+
+    def setUp(self):
+        super().setUp()
+        md_vaw = self.get_data_path('vaw.txt')
+        md_vaw = qiime2.Metadata.load(md_vaw)
+        self.md_vaw = md_vaw.get_column('Column')
+        table_vaw = self.get_data_path('vaw.qza')
+        self.table_vaw = qiime2.Artifact.load(table_vaw)
+        self.table_vaw, = feature_table.actions.filter_samples(
+            self.table_vaw, metadata=md_vaw)
+        imp = pd.read_csv(
+            self.get_data_path('vaw_importance.tsv'), sep='\t',
+            header=0, index_col=0)
+        self.imp = qiime2.Artifact.import_data('FeatureData[Importance]', imp)
+
+    def test_heatmap_default_feature_count_zero(self):
```

Comment:
is this test still working properly? with feature_count=0 it should be displaying all features... looks like there are 5 features in the input table, but only 4 in the heatmap? Why is one being dropped?

Reply:
That is a very good question, the answer to which is not immediately obvious to either myself or @ebolyen. Apologies for not noticing something like this earlier. To be perfectly honest my grasp of what these tests really do is shaky at best :laughing:. I'm looking into what might be causing this discrepancy now.

Reply:
see my review comment above for the cause... lack of a header line

Reply:
How did I not see that in your review or in the file 😂. I think it's time for me to be done for the week lol. Been doing too much stuff and getting not enough sleep.

Reply:
lol no worries! I think we are all ready for the weekend! Have a good one!